### DISCUSSION TITLE
aioespnow send method produces error on broadcast send (ValueError: Send timeout on synch.)

### DISCUSSION THREAD

**Discussion by garykuipers** [micropython/micropython#10004] [Using MicroPython] [answered]
Greetings: Relatively new to Micropython but knowledgeable about Python. Requesting assistance because I cannot find answers in my searches.

Here is the backtrace:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "main.py", line 70, in <module>
  File "uasyncio/core.py", line 1, in run
  File "uasyncio/core.py", line 1, in run_until_complete
  File "uasyncio/core.py", line 1, in run_until_complete
  File "main.py", line 66, in main
  File "uasyncio/funcs.py", line 1, in gather
  File "uasyncio/core.py", line 1, in run_until_complete
  File "main.py", line 49, in wait_peers
  File "comms.py", line 39, in get_peers
  File "comms.py", line 79, in register
  File "comms.py", line 72, in send_message
ValueError: Send timeout on synch.

 I am attaching the two files in the backtrace rather than presuming I know what to present. I appreciate your assistance.
[main.py.txt](https://github.com/micropython/micropython/files/10036088/main.py.txt) this is the main.py
[comms.py.txt](https://github.com/micropython/micropython/files/10036091/comms.py.txt) this is the comms.py

I found aioespnow.py at https://github.com/glenn20/micropython/blob/espnow-g20/ports/esp32/modules/aioespnow.py
it is attached.

[aioespnow.py.txt](https://github.com/micropython/micropython/files/10036582/aioespnow.py.txt)

The relevant section is:

```
22    async def asend(self, mac, msg=None, sync=None):
23        if msg is None:
24            msg, mac = mac, None  # If msg is None: swap mac and msg
25        yield uasyncio.core._io_queue.queue_write(self)
26        return self.send(mac, msg, sync)  # type: ignore
```

**Comment by karfas** [comment, marked as answer]
This is nothing special, I think.
You get a timeout when sending to `to_mac[0]` (which is most likely different from `to_mac[1]` printed) and you don't handle the resulting exception.
Maybe to_mac[0] is not started, disconnected, whatever ?

I don't have enough patience to analyze your whole program (this is your task, anyway).

```
    async def send_message(self, to_mac, msg: dict):
        msg_str = json.dumps(msg)
        print(f"Comms: send_message: to_mac={to_mac[1]} msg_str={msg_str}")
        await self.e.send(to_mac[0], msg_str.encode('ascii'))
```

**Reply by garykuipers** [reply]
That was a giant "DOH!" on my part. Thanks!

**Comment by glenn20** [comment]
@garykuipers  Thanks for the report. As you can see from [esp_espnow.c#L504](https://github.com/glenn20/micropython/blob/espnow-g20/ports/esp32/esp_espnow.c#L504) this exception "should not happen". The ESP-NOW software stack is meant to invoke the send_cb() function (which updates self->tx_responses) when a sent packet is acknowledged by the peer OR after no response has been received (which is usually invoked within 25ms or so). I have never seen this happen, but it is interesting to know that it can.

So, yes, as suggested by @karfas , best to just trap the exception and continue. I'll review my code to see if there is a better way to handle this now that I know it CAN occur or if there is some logic error on my part causing it to occur.

Oh - and if this is happening rarely, it should have little impact on your application, BUT if it happen frequently then you will be carrying the burden of the 250ms timeout whenever this occurs. If that is the case, give me some more details about your device and I will try to test.
